protocols/connection-limit/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+use libp2p_swarm::{NetworkBehaviour, NetworkBehaviourAction, PollParameters, ReviewDenied};
+use std::error::Error;
+use std::fmt;
+use std::task::{Context, Poll};
+
+pub struct Behaviour {
+    counters: ConnectionCounters,
+}
+
+impl Behaviour {
+    pub fn new(limits: ConnectionLimits) -> Self {
+        Behaviour {
+            counters: ConnectionCounters::new(limits),
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = DummyConnectionHandler;
+    type OutEvent = ();
+
+    fn new_handler(&mut self) -> Self::ConnectionHandler {
+        Default::default()
+    }
+
+    fn review_pending_connection(
+        &mut self,
+        _peer_id: Option<PeerId>,
+        // TODO: Maybe an iterator is better?
+        _addresses: &[Multiaddr],
+        endpoint: Endpoint,
+    ) -> Result<(), ReviewDenied> {
+        match endpoint {
+            Endpoint::Dialer => self.counters.check_max_pending_outgoing(),
+            Endpoint::Listener => self.counters.check_max_pending_incoming(),
+        }
+        .map_err(|e| ReviewDenied::Error(e.into()))
+    }
+
+    fn review_established_connection(
+        &mut self,
+        _peer_id: PeerId,
+        // TODO: Maybe an iterator is better?
+        endpoint: &ConnectedPoint,
+    ) -> Result<(), ReviewDenied> {
+        self.counters
+            .check_max_established(endpoint)
+            .map_err(|e| ReviewDenied::Error(e.into()))
+    }
+
+    fn inject_connection_pending(
+        &mut self,
+        _peer_id: Option<PeerId>,
+        _connection_id: ConnectionId,
+        endpoint: Endpoint,
+    ) {
+        self.counters.inc_pending(endpoint)
+    }
+
+    fn inject_connection_established(
+        &mut self,
+        _peer_id: &PeerId,
+        _connection_id: &ConnectionId,
+        endpoint: &ConnectedPoint,
+        _failed_addresses: Option<&Vec<Multiaddr>>,
+        _other_established: usize,
+    ) {
+        self.counters.dec_pending(endpoint);
+        self.counters.inc_established(endpoint)
+    }
+
+    fn inject_connection_closed(
+        &mut self,
+        _: &PeerId,
+        _: &ConnectionId,
+        endpoint: &ConnectedPoint,
+        _: <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
+        _remaining_established: usize,
+    ) {
+        self.counters.dec_established(endpoint)
+    }
+
+    fn inject_event(
+        &mut self,
+        _peer_id: PeerId,
+        _connection: ConnectionId,
+        _event: <<Self::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    ) {
+    }
+
+    fn poll(
+        &mut self,
+        _cx: &mut Context<'_>,
+        _params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ConnectionHandler>> {
+        Poll::Pending
+    }
+}
+
+/// Network connection information.
+#[derive(Debug, Clone)]
+pub struct ConnectionCounters {
```

Comment:
2 things:

1. Do we need this indirection?
2. If yes, does it have to be `pub`?

Reply:
Does not. Good catch. Just copied this from the previous connection limiting mechanism.